### ISSUE TITLE
Unstable rule selection by the GHOST action selector

### ISSUE THREAD

**Issue by joejnke** [singnet/opencog-tutorials#15]
The GHOST action selector makes unpredictable rule selection between two same rules but different in their goal strength `(ghost-parse "#goal: (goal1=0.7) u: (hi) hello there ^keep")`  and `(ghost-parse "#goal: (goal1=0.5) u: (hi) hi there ^keep")`. Running the code `(test-ghost "hi")` gives sometimes `((WordNode "hi")
 (WordNode "there")
)` and other times `((WordNode "hello")
 (WordNode "there")
)`.
The expected response was expected to be of the rule with the highest goal strength, that is `((WordNode "hello")
 (WordNode "there")
)`

**Comment by simonmekit** [comment]
There are also inconsistencies in output of rule goal strength with `ghost-rule-tv` command. 

(FYI: tested without starting cog-server and without using ecan based rules)

scheme@(guile-user)> (ghost-parse "#goal: (goal1=0.4) u: lbl2(hi) hi there ^keep")
scheme@(guile-user)> (ghost-parse "#goal: (goal1=0.8) u: lbl2(hi) hello there ^keep")
scheme@(guile-user)> (test-ghost "hi")
[2018-10-31 09:43:29:500] [WARN] No configuration file was loaded! Param=STARTING_STI_FUNDS
[2018-10-31 09:43:29:500] [WARN] No configuration file was loaded! Param=STARTING_LTI_FUNDS
[2018-10-31 09:43:29:500] [WARN] No configuration file was loaded! Param=STI_FUNDS_BUFFER
[2018-10-31 09:43:29:500] [WARN] No configuration file was loaded! Param=LTI_FUNDS_BUFFER
[2018-10-31 09:43:29:500] [WARN] No configuration file was loaded! Param=TARGET_LTI_FUNDS
[2018-10-31 09:43:29:500] [WARN] No configuration file was loaded! Param=TARGET_STI_FUNDS
[2018-10-31 09:43:29:500] [WARN] No configuration file was loaded! Param=ECAN_STARTING_ATOM_STI_WAGE
[2018-10-31 09:43:29:500] [WARN] No configuration file was loaded! Param=ECAN_STARTING_ATOM_LTI_WAGE
[2018-10-31 09:43:29:500] [WARN] No configuration file was loaded! Param=ECAN_MAX_AF_SIZE
[2018-10-31 09:43:29:505] [INFO] [say] (hello there)
$1 = ((WordNode "hello")
 (WordNode "there")
)
Action selection seems correct in this case, but running `ghost-rule-tv "lbl2"` returned the other value. 
scheme@(guile-user)> (ghost-rule-tv "lbl2")
$2 = (stv 0.400000 0.900000)

The top level goal strength is not recognized by the same command:

scheme@(guile-user)> (ghost-parse "goal: (please_user=0.8)")
scheme@(guile-user)> (ghost-parse "u: lbl1 (hello) Hello wonderful human")
[2018-10-30 12:38:31:960] [WARN] [GHOST] Did you forget to link a goal to the rule?
scheme@(guile-user)> (ghost-rule-tv "lbl1")
$2 = (stv 0.900000 0.900000)

**Comment by leungmanhin** [comment]
@joejnke Action selection is probabilistic, so not always choosing the one with the highest weight is the expected behavior here.

**Comment by leungmanhin** [comment]
@simonmekit Again, put everything in a text file and use `ghost-parse-files` to parse rules. If you want to call `ghost-parse` repeatedly, don't forget to put all the related goals/rules etc in one call, i.e. `(ghost-parse "#goal: (please_user=0.8) u: lbl1 (hello) Hello wonderful human")` in this case.

See also: https://github.com/singnet/opencog/issues/30

**Comment by simonmekit** [comment]
yes, indeed using `ghost-parse-file` makes stable action selection for same rule with different goal strength. 
The [Ghost tutorial ](https://github.com/singnet/opencog-tutorials/blob/master/opencog_tutorials/GHOST/ghost-tutorial.ipynb) could be updated with `ghost-parse-file` in addition to  `ghost-parse` to align with subsequent test process.

**Comment by simonmekit** [comment]
@leungmanhin as you mentioned rule-level goal also works fine with `ghost-parse-file` and  `ghost-rule-tv`  commands. However, it seems the concept of top-level goal described in the [tutorial](https://github.com/singnet/opencog-tutorials/blob/master/opencog_tutorials/GHOST/ghost-tutorial.ipynb) is not properly working. 

"ghostfile.ghost" contains these subsequent lines :
#goal: (goal1=0.4) u: lbl2(hi) hi there ^keep
#goal: (goal1=0.8) u: lbl2(hi) hello there ^keep
goal: (please_user=0.8)
u: lbl1(hello) Hello wonderful human
u: lbl3(Dr. Watson) hello everyone ^keep

scheme@(guile-user)> (ghost-rule-tv "lbl2")
$1 = (stv 0.800000 0.900000)
scheme@(guile-user)> (ghost-rule-tv "lbl1")
$2 = (stv 0.800000 0.900000)
scheme@(guile-user)> (ghost-rule-tv "lbl3")
$3 = (stv 0.800000 0.900000)

**Comment by leungmanhin** [comment]
> 
> "ghostfile.ghost" contains these subsequent lines :
> #goal: (goal1=0.4) u: lbl2(hi) hi there ^keep
> #goal: (goal1=0.8) u: lbl2(hi) hello there ^keep
> goal: (please_user=0.8)
> u: lbl1(hello) Hello wonderful human
> u: lbl3(Dr. Watson) hello everyone ^keep
> 
> scheme@(guile-user)> (ghost-rule-tv "lbl2")
> $1 = (stv 0.800000 0.900000)
> scheme@(guile-user)> (ghost-rule-tv "lbl1")
> $2 = (stv 0.800000 0.900000)
> scheme@(guile-user)> (ghost-rule-tv "lbl3")
> $3 = (stv 0.800000 0.900000)

@simonmekit Sorry I don't see any problem in the above example... What is the expected result? Or are you just talking about the ones in the tutorial?

**Comment by simonmekit** [comment]
yes, I was running tests to make sure that the tutorial complies with current updates and prepare some test-suite. Top level goal works fine with `ghost-parse-file`, sorry @leungmanhin for confusion!

**Comment by leungmanhin** [comment]
@simonmekit no problem, thanks for doing all these!
